[contributing] Recommend static C functions for common initialization

stephanemoore opened the pull request:
Recommend static C functions for common initialization instead of private instance methods with collision-resistant names.

stephanemoore committed:
Recommend static C functions for common initialization

Recommend static C functions for common initialization instead of private instance methods with collision-resistant names.

randallli reviewed (approved):
LGTM.
I added some others who may have stronger opinions.

originally came up in cl/147566240.

ajsecord reviewed (approved):
👍  Please file a bug to change the current usage in MDC-iOS.

ianegordon commented:
I'm in favor of the static scoping of the function, but I'm concerned this might be surprising to clients.
I don't know that I've seen a C function take self as a parameter.
Let me do some digging.

ianegordon reviewed (changes requested):
Looking for best / common practices.

willlarche commented:
That was the same concern I had. It's a little avant-garde for something so
mundane.

On Mon, May 1, 2017 at 11:51 AM ianegordon <notifications@github.com> wrote:

> I'm in favor of the static scoping of the function, but I'm concerned this
> might be surprising to clients.
> I don't know that I've seen a C function take self as a parameter.
> Let me do some digging.
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/material-components/material-components-ios/pull/1346#issuecomment-298359670>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABNm5RR2LLmNSMWPHAwFf2KlwSZepCeQks5r1f-agaJpZM4NMGyP>
> .
>

ianegordon commented:
Initial thinking is to decline based on preferring Objective-C methods to C functions.
Discussing offline.

stephanemoore commented:
Regarding C functions with parameters named self: it's not that uncommon. Essentially all Objective-C methods are functions with two hidden parameters "self" and "_cmd". Naming a C function with a similar structure helps identify the function as an operation on an Objective-C object.

You can see examples of this in Apple open source code and documentation:
https://developer.apple.com/reference/objectivec/1456712-objc_msgsend?language=objc
https://opensource.apple.com/source/objc4/objc4-709/runtime/objc-runtime-new.mm.auto.html

You can also see examples in Google open source code:
https://github.com/google/protobuf/blob/master/objectivec/GPBMessage.m

stephanemoore commented:
Regarding the preference for Objective-C methods: are you worried that contributors will find static C functions unfamiliar or alienating?

If you look through the UIKit symbol table you will find many examples indicative of C functions used to consolidate initialization logic. With Xcode 8.3 I ran the following command on UIKit:
```
$ nm /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/Frameworks/UIKit.framework/UIKit -s __TEXT __text | grep -i -e "common.*init"
```
You will observe C function symbols including UIViewCommonInitWithFrame, UILabelCommonInit, and _UISwitchCommonInit. These all seem like clear examples of C functions used to consolidate initialization logic.

An important observation from above is that said command only reveals the common initialization C functions that have external linkage. Any common initialization C functions with internal linkage are more difficult to detect.

ianegordon commented:
As you noted, I am concerned contributors would find static C functions unfamiliar or alienating.

self->_titleLabel.text = @"Default Title"; // Seems a bit unfamiliar to have the self->

The symbol table is pretty interesting.

UILabel appears to have both a commonInit method and function:
00000000002aa1d6 t -[UILabel _commonInit]
00000000002aa396 t _UILabelCommonInit

Many classes have _commonInit methods, while others are named with the class, such as [UISegment _commonSegmentInit].

Let me do a bit more digging.

stephanemoore committed:
Update example for proper ivar access inheritance

Move CommonInit into the @implementation block because C functions declared inside the @implementation block inherit the access privileges of the class. Having the C function outside of the @implementation block is possible but requires other changes.

ianegordon commented:
To close the loop.  We agreed to move forward with the specifically-named class methods of commonMDCComponentInit.

Many Thanks for the PR and the nm | grep fu.
